src/FsAutoComplete/LSP.Preview.fs

```diff
@@ -0,0 +1,145 @@
+    /// and it shows a context menu with further code navigation commands.
+    ///
+    /// Depending on the client capability `inlayHint.resolveSupport` clients
+    /// might resolve this property late using the resolve request.
+    Location: Location option
+    /// An optional command for this label part.
+    ///
+    /// Depending on the client capability `inlayHint.resolveSupport` clients
+    /// might resolve this property late using the resolve request.
+    Command: Command option }
+
+[<ErasedUnion>]
+[<RequireQualifiedAccess>]
+type InlayHintLabel =
+  | String of string
+  | Parts of InlayHintLabelPart[]
+
+/// Inlay hint information.
+type InlayHint<'Data> =
+  { /// The position of this hint.
+    Position: Position
+    /// The label of this hint. A human readable string or an array of
+    /// InlayHintLabelPart label parts.
+    ///
+    /// *Note* that neither the string nor the label part can be empty.
+    Label: InlayHintLabel
+    /// he kind of this hint. Can be omitted in which case the client
+    /// should fall back to a reasonable default.
+    Kind: InlayHintKind option
+    /// Optional text edits that are performed when accepting this inlay hint.
+    ///
+    /// *Note* that edits are expected to change the document so that the inlay
+    /// hint (or its nearest variant) is now part of the document and the inlay
+    /// hint itself is now obsolete.
+    ///
+    /// Depending on the client capability `inlayHint.resolveSupport` clients
+    /// might resolve this property late using the resolve request.
+    TextEdits: TextEdit[] option
+    /// The tooltip text when you hover over this item.
+    ///
+    /// Depending on the client capability `inlayHint.resolveSupport` clients
+    /// might resolve this property late using the resolve request.
+    Tooltip: InlayHintTooltip option
+    /// Render padding before the hint.
+    ///
+    /// Note: Padding should use the editor's background color, not the
+    /// background color of the hint itself. That means padding can be used
+    /// to visually align/separate an inlay hint.
+    PaddingLeft: bool option
+    /// Render padding after the hint.
+    ///
+    /// Note: Padding should use the editor's background color, not the
+    /// background color of the hint itself. That means padding can be used
+    /// to visually align/separate an inlay hint.
+    PaddingRight: bool option
+
+    /// A data entry field that is preserved on a inlay hint between
+    /// a `textDocument/inlayHint` and a `inlayHint/resolve` request.
+    ///
+    /// Note: In LSP specs: of type `LSPAny`:
+    /// ```typescript
+    /// export type LSPAny = LSPObject | LSPArray | string | integer | uinteger | decimal | boolean | null;
+    /// export type LSPObject = { [key: string]: LSPAny };
+    /// export type LSPArray = LSPAny[];
+    /// ```
+    /// -> `'Data` must adhere to specs
+    Data: 'Data option }
```

Comment:
`Data` is of type [`LSPAny`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#lspAny):
```typescript
export type LSPAny = LSPObject | LSPArray | string | integer | uinteger |
	decimal | boolean | null;
export type LSPObject = { [key: string]: LSPAny };
export type LSPArray = LSPAny[];
```
Not really representable in F#  
-> I decided to use a Generic instead

But then of course correctness is in user hand and not type checked any more. I still think it's way easier to use than anything else.

Reply:
LSPAny looks a lot like JToken. In other cases like this we've used JToken/JToken[] as appropriate, because the core need is a JSON-serializable data blob.

Reply:
Yes -- but it's so horrible to use in F# (compared to a normal record)....

(though I think it can be used with additional (de-)serialization -> same as generic -- but with a step between)

Reply:
oh, I see - you're less concerned with 100% representing the API contract, because you're also trying to make sure people use the API correctly: by specifying a generic type, you let the JSON-RPC handler deserialize it consistently to prevent data out-of-sync errors.

Reply:
I wouldn't call it "out-of-sync error" -- it requires an erroneous or malicious client to actively change `Data` (`Data` is just a short-lived note from Server to itself which the client just has to preserve).  
But generic `Data` places the responsibility for correctness (and error handling) of `Data` on LSP side (-> `Data` should be preserved -- otherwise violation of LSP contract (vs. violation of FSAC contract; for [example](https://github.com/fsharp/FsAutoComplete/blob/88c747d316a5868e1b9be08ab72bfe10526ec055/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L767-L792) with valid `InitializeParams` (LSP), but invalid `FSharpConfigDto` (FSAC) -> FSAC responsibility to handle error))  


<br/>

But when making `Data` generic, I only considered a Server implementation. In a client implementation, the client shouldn't care about `Data` at all (besides preserving it). But with a generic the client MUST specify some type -- the opposite of "not caring"... `JToken` on the other hand is just that: some json I can just ignore.  
-> I'll change `InlayHint` to use `JToken` instead